### PR TITLE
#4 | Add products and users tables

### PR DESCRIPTION
**Pull request by xonism**

### CODE REVIEW

File: internship-warehouse/src/main/java/com/twoday/internshipwarehouse/controllers/ProductController.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    public ProductController(ProductService productService) {
+        this.productService = productService;
+    }
+
+    @GetMapping
+    public ResponseEntity<List<ProductDTO>> getAllProducts() {
+        List<Product> products = productService.getAll();
+        return new ResponseEntity<>(
+                products.stream().map(this::convertToDTO).toList(),
```
Comment (ArnasPetraitis) [inline internship-warehouse/src/main/java/com/twoday/internshipwarehouse/controllers/ProductController.java]: streams are more readable when actions are line-breaked.  I would also separated this stream to a variable, e.g. you dont use product list so maping could be done there. e.g.
```
List<ProductDTO> products = productService.getAll()
           .stream()
           .map(this::convertToDTO)
           .toList();
return ResponseEntity.ok().body(products);
```

File: internship-warehouse/src/main/java/com/twoday/internshipwarehouse/security/CustomAuthenticationProvider.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+
+@Component
+public class CustomAuthenticationProvider implements AuthenticationProvider {
```
Comment (ArnasPetraitis) [inline internship-warehouse/src/main/java/com/twoday/internshipwarehouse/security/CustomAuthenticationProvider.java]: This is ok, but is to complicated for current task. Try implement own UserDetailsService and map User to UserDetails there

File: internship-warehouse/src/main/java/com/twoday/internshipwarehouse/services/ProductService.java [hunk 3]
```diff
@@ -1,35 +1,35 @@
...
+        return productRepository.findAll();
     }
 
     public Product processSale(ProductSellRequest productSellRequest) {
         if (productSellRequest.quantity() <= 0) {
             throw new InvalidValueException(Constants.QUANTITY);
         }
 
-        Product product = warehouseRepository.findById(productSellRequest.id());
+        Product product = productRepository.findById(productSellRequest.id()).orElse(null);
```
Comment (ArnasPetraitis) [inline internship-warehouse/src/main/java/com/twoday/internshipwarehouse/services/ProductService.java]: `.orElse(null)` is bad practice. especially when you do  `if (product == null)` on next line. 
```
Product product = productRepository.findById(productSellRequest.id())
     .orElseTrow(()->new ProductNotFoundByIdException(productSellRequest.id());
```

**Review by ArnasPetraitis — CHANGES_REQUESTED** [review]

File: internship-warehouse/src/main/java/com/twoday/internshipwarehouse/services/ProductService.java [hunk 4]
```diff
@@ -29,7 +29,8 @@ public Product processSale(ProductSellRequest productSellRequest) {
             throw new InvalidValueException(Constants.QUANTITY);
         }
 
-        Product product = productRepository.findById(productSellRequest.id()).orElse(null);
+        Product product = productRepository.findById(productSellRequest.id())
+                .orElseThrow(() -> new ProductNotFoundByIdException(productSellRequest.id()));
 
         if (product == null) {
```
Comment (ArnasPetraitis) [inline internship-warehouse/src/main/java/com/twoday/internshipwarehouse/services/ProductService.java]: should be removed?

**Review by ArnasPetraitis — CHANGES_REQUESTED** [review]

File: internship-warehouse/src/main/java/com/twoday/internshipwarehouse/services/ProductService.java (hunk 4 above)
Comment (xonism) [reply internship-warehouse/src/main/java/com/twoday/internshipwarehouse/services/ProductService.java]: Accidentally didn't commit it 😅 Should be fixed now

**Review by ArnasPetraitis — APPROVED** [review]

**Pull request merged**